### DISCUSSION TITLE
Fail parser immediately (skip other choices)

### DISCUSSION THREAD

**Discussion by zaynetro** [zesterer/chumsky#522] [Q&A] [answered]
Let's say that I have this parser: it either parses a function definition or an expression.

```rust
let fn_ = keyword("fn").padded()
    .ignore_then(ident_)
    .then(args)
    .then(fn_body);

choice((
    fn_,
    expr,
))
```

How do I prevent my parser from falling back to expression parsing after it has encountered a `fn` keyword and then failed?

Nom has a [`cut`](https://docs.rs/nom/latest/nom/combinator/fn.cut.html) combinator that fails the parser completely. Is there a similar thing in chumsky?

Why do I need this? If there was a problem with function arguments I want to see the error there instead of an error due it my parser failing to parse an expression.

**Comment by zesterer** [comment, marked as answer]
Hello,

> If there was a problem with function arguments I want to see the error there instead of an error due it my parser failing to parse an expression.

Chumsky will *automatically* prioritise error branches that make the most progress into the input. This means that if the parser sees `fn <malformed>`, the error you will see will be that generated by the `fn_` parser (unless, of course, the `expr` parser contains some syntax rule that can parse the `fn` too).

The *exact* mechanism for this is kept as an implementation detail (there are some future changes we might be able to do to improve error prioritisation even further), but the parser will already exhibit the behaviour you're looking for without any extra combinators :)

**Reply by zaynetro** [reply]
oh wow. That's pretty cool! I think you can put this answer into [error_and_recovery](https://github.com/zesterer/chumsky/blob/main/guide/error_and_recovery.md) guide.

I guess I need to play around more with error cases to see how it works in practice.
